src/nodes.js

```diff
@@ -143,10 +154,44 @@ var registerChild = function(parent, key, child) {
 };
 
 
+/**
+ * Marks the key map for a given Element as invalid. It will be cleaned the
+ * next time it is needed.
+ * @param {!Element} el
+ */
+var invalidateKeyMap = function(el) {
+  getKeyMap(el).isValid = false;
+};
+
+
+/**
+ * Cleans a key map for a given Element, removing any nodes that are no longer
+ * in the DOM.
+ * @param {!Element} el
+ */
+var cleanKeyMap = function(el) {
+  var keyMap = getData(el).keyMap;
+
+  if (!keyMap || keyMap.isValid) {
+    return;
+  }
+
+  for (var key in keyMap) {
+    if (!keyMap[key].parentNode) {
+      delete keyMap[key];
```

Comment:
> Holding on to a single node is okay, but you could end up holding on to very large subtrees.

Are you talking about the bookkeeping approach? If so, the child is [nulled out](https://github.com/jridgewell/incremental-dom/commit/b776a9a#diff-db8f78209fc0a7d78756881ed982a3c3R151), preventing the memory leak. All the bookkeeping does is prevent the `keyMap` from having _more than_ 20 empty slots. So if only once keyed child is removed, we have something like:

```js
keyMap = {
  child1: <div>,
  child2: null,
  child3: <div>,
  //...
}
```